cpp/src/sortedmerge/sortedmerge.cu

```diff
@@ -0,0 +1,289 @@
+    auto right_begin_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(right_side, right_indices));
+
+    auto left_end_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(left_side + left_size, left_indices + left_size));
+    auto right_end_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(right_side + right_size, right_indices + right_size));
+
+    rmm::device_vector<gdf_size_type> outputTableIndices(total_size);
+    rmm::device_vector<gdf_size_type> outputRowIndices(total_size);
+    auto output_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(outputTableIndices.begin(), outputRowIndices.begin()));
+
+    bool nullable = left_table.has_nulls() || right_table.has_nulls();
+    if (nullable){
+        auto ineq_op = row_inequality_comparator<true>(right_table, left_table, nulls_are_smallest, asc_desc.data().get()); 
+        thrust::merge(rmm::exec_policy(stream)->on(stream),
+                    left_begin_zip_iterator,
+                    left_end_zip_iterator,
+                    right_begin_zip_iterator,
+                    right_end_zip_iterator,
+                    output_zip_iterator,
+                    [=] __device__ (thrust::tuple<gdf_size_type, gdf_size_type> const & tupleA,
+                                    thrust::tuple<gdf_size_type, gdf_size_type> const & tupleB) {
+                        const gdf_size_type left_row = thrust::get<0>(tupleA) == LEFT_TABLE ? thrust::get<1>(tupleA) : thrust::get<1>(tupleB);
+                        const gdf_size_type right_row = thrust::get<0>(tupleA) == RIGHT_TABLE ? thrust::get<1>(tupleA) : thrust::get<1>(tupleB);
+                        return ineq_op(right_row, left_row);
+                    });			        
+    } else {
+        auto ineq_op = row_inequality_comparator<false>(right_table, left_table, nulls_are_smallest, asc_desc.data().get()); 
+        thrust::merge(rmm::exec_policy(stream)->on(stream),
+                    left_begin_zip_iterator,
+                    left_end_zip_iterator,
+                    right_begin_zip_iterator,
+                    right_end_zip_iterator,
+                    output_zip_iterator,
+                    [=] __device__ (thrust::tuple<gdf_size_type, gdf_size_type> const & tupleA,
+                                    thrust::tuple<gdf_size_type, gdf_size_type> const & tupleB) {
+                        const gdf_size_type left_row = thrust::get<0>(tupleA) == LEFT_TABLE ? thrust::get<1>(tupleA) : thrust::get<1>(tupleB);
+                        const gdf_size_type right_row = thrust::get<0>(tupleA) == RIGHT_TABLE ? thrust::get<1>(tupleA) : thrust::get<1>(tupleB);
+                        return ineq_op(right_row, left_row);
+                    });					        
+    }
+
+    CHECK_STREAM(stream);
+
+    return std::make_pair(outputTableIndices, outputRowIndices);
+}
+
+} // namespace
+
+namespace cudf {
+namespace detail {
+
+table sorted_merge(table const& left_table,
```

Comment:
To make later stream support easier, this `detail::sorted_merge` should take a stream that is defaulted to zero. Then use the stream for all execution, allocation, and copies inside the function.